ext/network.go

```diff
@@ -0,0 +1,488 @@
+		),
+		cel.Function(isUnspecifiedFunc,
+			cel.MemberOverload("ip_isUnspecified", []*cel.Type{networkIPType}, cel.BoolType,
+				cel.UnaryBinding(netIPIsUnspecified)),
+		),
+		cel.Function(isLinkLocalMcastFunc,
+			cel.MemberOverload("ip_isLinkLocalMulticast", []*cel.Type{networkIPType}, cel.BoolType,
+				cel.UnaryBinding(netIPIsLinkLocalMulticast)),
+		),
+		cel.Function(isLinkLocalUcastFunc,
+			cel.MemberOverload("ip_isLinkLocalUnicast", []*cel.Type{networkIPType}, cel.BoolType,
+				cel.UnaryBinding(netIPIsLinkLocalUnicast)),
+		),
+		cel.Function(containsIPFunc,
+			cel.MemberOverload("cidr_containsIP_ip", []*cel.Type{networkCIDRType, networkIPType}, cel.BoolType,
+				cel.BinaryBinding(netCIDRContainsIP)),
+			cel.MemberOverload("cidr_containsIP_string", []*cel.Type{networkCIDRType, cel.StringType}, cel.BoolType,
+				cel.BinaryBinding(netCIDRContainsIPString)),
+		),
+		cel.Function(containsCIDRFunc,
+			cel.MemberOverload("cidr_containsCIDR_cidr", []*cel.Type{networkCIDRType, networkCIDRType}, cel.BoolType,
+				cel.BinaryBinding(netCIDRContainsCIDR)),
+			cel.MemberOverload("cidr_containsCIDR_string", []*cel.Type{networkCIDRType, cel.StringType}, cel.BoolType,
+				cel.BinaryBinding(netCIDRContainsCIDRString)),
+		),
+		cel.Function(maskedFunc,
+			cel.MemberOverload("cidr_masked", []*cel.Type{networkCIDRType}, networkCIDRType,
+				cel.UnaryBinding(netCIDRMasked)),
+		),
+		cel.Function(prefixLengthFunc,
+			cel.MemberOverload("cidr_prefixLength", []*cel.Type{networkCIDRType}, cel.IntType,
+				cel.UnaryBinding(netCIDRPrefixLength)),
+		),
+		cel.Function(ipFromCIDRFunc,
+			cel.MemberOverload("cidr_ip", []*cel.Type{networkCIDRType}, networkIPType,
+				cel.UnaryBinding(netCIDRIP)),
+		),
+	}
+}
+
+func (*networkLib) ProgramOptions() []cel.ProgramOption {
+	return []cel.ProgramOption{}
+}
+
+// networkAdapter adapts netip types while preserving existing adapters.
+type networkAdapter struct {
+	types.Adapter
+}
+
+func (a *networkAdapter) NativeToValue(value any) ref.Val {
+	switch v := value.(type) {
+	case netip.Addr:
+		return IP{Addr: v}
+	case netip.Prefix:
+		return CIDR{Prefix: v}
+	}
+	// Delegate to the wrapped adapter (e.g., Protobuf adapter)
+	return a.Adapter.NativeToValue(value)
+}
+
+// --- Implementation Logic ---
+
+func parseIPAddr(raw string) (netip.Addr, error) {
+	addr, err := netip.ParseAddr(raw)
+	if err != nil {
+		return netip.Addr{}, fmt.Errorf("IP Address %q parse error: %v", raw, err)
+	}
+	if addr.Zone() != "" {
+		return netip.Addr{}, fmt.Errorf("IP address %q with zone value is not allowed", raw)
+	}
+	if addr.Is4In6() {
+		return netip.Addr{}, fmt.Errorf("IPv4-mapped IPv6 address %q is not allowed", raw)
+	}
+	return addr, nil
+}
+
+func netIsIP(val ref.Val) ref.Val {
+	s := val.(types.String)
+	_, err := parseIPAddr(string(s))
+	return types.Bool(err == nil)
+}
+
+func netIsCIDR(val ref.Val) ref.Val {
+	s := val.(types.String)
+	_, err := netip.ParsePrefix(string(s))
```

Comment:
Can you make sure the function implementations are being done consistently with K8s? There are some specific constraints that K8s enforces which I'm not seeing here. Might be way to copy the function implementation and make the main edit the removal of the unnecessary function guards.

https://github.com/kubernetes/kubernetes/blob/5bcb7599736327cd8c6d23e398002354a6e40f68/staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go#L281